Cannot read, edit or delete a file when the guest link is deleted

I created a guest link, just for testing something.
I uploaded a file and deleted the link.
Since I deleted the link, I have not been able to read/open, edit or delete the file, and I got an error message: "_failed to retrieve entry_"

Screenshot:
![DuckDuckGo_2023-10-01_11-36-36](https://github.com/mtlynch/picoshare/assets/30030291/e925b8df-d26f-4bb6-b378-901b414cfe12)

System Info (If necessary):
Synology 718+
DSM 7.2.1-69057
Docker

Thanks for reporting this! I can reproduce this. The error message in the logs is:

```text
2023/10/01 15:29:53 download.go:32: error retrieving entry with id 4FgxWPDuqg: sql: Scan error on column index 6, name "guest_link_id": converting NULL to string is unsupported
```

It looks like we're not handling it properly when the guest link disappears but the entry remains.

I have a fix that seems to work, but I need to add some tests as well.

https://github.com/mtlynch/picoshare/pull/502

--- a/e2e/guest-links.spec.ts
+++ b/e2e/guest-links.spec.ts
@@ -1,5 +1,6 @@
 import { test, expect } from "@playwright/test";
 import { login } from "./helpers/login.js";
+import { readDbTokenCookie } from "./helpers/db.js";
 
 const labelColumn = 0;
 
@@ -62,4 +63,75 @@ test("creates a guest link and uploads a file as a guest", async ({ page }) => {
 
   await expect(page.locator("h1")).toContainText("Guest Link Inactive");
   await expect(page.locator(".file-input")).toHaveCount(0);
+});
+
+test("files uploaded through guest link remain accessible after guest link is deleted", async ({
+  page,
+  browser,
+}) => {
+  await login(page);
+
+  await page.getByRole("menuitem", { name: "Guest Links" }).click();
+
+  await page.getByRole("button", { name: "Create new" }).click();
+
+  await expect(page).toHaveURL("/guest-links/new");
+  await page.locator("#label").fill("I'll be deleted soon");
+  await page.locator("#file-upload-limit").fill("1");
+  await page.getByRole("button", { name: "Create" }).click();
+
+  await expect(page).toHaveURL("/guest-links");
+  const guestLinkRow = await page
+    .getByRole("row")
+    .filter({ hasText: "I'll be deleted soon" });
+  await expect(guestLinkRow).toBeVisible();
+
+  // Save the route to the guest link URL so that we can return to it later.
+  const guestLinkRouteValue = await guestLinkRow
+    .getByRole("cell")
+    .nth(labelColumn)
+    .getByRole("link")
+    .getAttribute("href");
+  expect(guestLinkRouteValue).not.toBeNull();
+  const guestLinkRoute = String(guestLinkRouteValue);
+
+  {
+    const guestContext = await browser.newContext();
+
+    // Share database across users.
+    await guestContext.addCookies([
+      readDbTokenCookie(await page.context().cookies()),
+    ]);
+
+    const guestPage = await guestContext.newPage();
+
+    await guestPage.goto(guestLinkRoute);
+    await guestPage.locator(".file-input").setInputFiles([
+      {
+        name: "guest-upload2.txt",
+        mimeType: "text/plain",
+        buffer: Buffer.from("uploaded by a guest user"),
+      },
+    ]);
+    await expect(guestPage.locator("#upload-result .message-body")).toHaveText(
+      "Upload complete!"
+    );
+  }
+
+  await guestLinkRow.getByRole("button", { name: "Delete" }).click();
+
+  await page.getByRole("menuitem", { name: "Files" }).click();
+
+  const filenameColumn = 0;
+  //await page.getByRole("row").filter({ hasText: "guest-upload2.txt" }).click();
+
+  await page
+    .getByRole("row")
+    .filter({ hasText: "guest-upload2.txt" })
+    .getByRole("cell")
+    .nth(filenameColumn)
+    .getByRole("link")
+    .click();
+
+  await expect(page.locator("body")).toHaveText("uploaded by a guest user");
 });

--- a/e2e/guest-links.spec.ts
+++ b/e2e/guest-links.spec.ts
@@ -123,8 +123,6 @@ test("files uploaded through guest link remain accessible after guest link is de
   await page.getByRole("menuitem", { name: "Files" }).click();
 
   const filenameColumn = 0;
-  //await page.getByRole("row").filter({ hasText: "guest-upload2.txt" }).click();
-
   await page
     .getByRole("row")
     .filter({ hasText: "guest-upload2.txt" })

--- a/e2e/helpers/db.js
+++ b/e2e/helpers/db.js
@@ -1,0 +1,8 @@
+export function readDbTokenCookie(cookies) {
+  for (const cookie of cookies) {
+    if (cookie.name === "db-token") {
+      return cookie;
+    }
+  }
+  return undefined;
+}

--- a/store/sqlite/entries.go
+++ b/store/sqlite/entries.go
@@ -98,7 +98,7 @@ func (d DB) GetEntryMetadata(id picoshare.EntryID) (picoshare.UploadMetadata, er
 	var uploadTimeRaw string
 	var expirationTimeRaw string
 	var fileSize uint64
-	var guestLinkID picoshare.GuestLinkID
+	var guestLinkID *picoshare.GuestLinkID
 	err := d.ctx.QueryRow(`
 	SELECT
 		entries.filename AS filename,
@@ -129,8 +129,8 @@ func (d DB) GetEntryMetadata(id picoshare.EntryID) (picoshare.UploadMetadata, er
 	}
 
 	var guestLink picoshare.GuestLink
-	if !guestLinkID.Empty() {
-		guestLink, err = d.GetGuestLink(guestLinkID)
+	if guestLinkID != nil && !guestLinkID.Empty() {
+		guestLink, err = d.GetGuestLink(*guestLinkID)
 		if err != nil {
 			return picoshare.UploadMetadata{}, err
 		}

